cmd/kubeadm/app/phases/kubelet/config.go

```diff
@@ -120,27 +123,91 @@ func createConfigMapRBACRules(client clientset.Interface, k8sVersion *version.Ve
+// DownloadAndMergeConfig downloads the kubelet configuration from a ConfigMap, merge with the local one and writes it to disk.
+// Used at "kubeadm upgrade" time
+func DownloadAndMergeConfig(client clientset.Interface, kubeletVersion *version.Version, kubeletDir string) error {
+
+	kubeletConfig, err := fetchConfigFromConfigMap(client, kubeletVersion, kubeletDir)
+	if err != nil {
+		return err
+	}
+
+	cfg, err := getConfigFromFile(kubeletDir)
+	if err != nil {
+		klog.Warningf("getConfigFromFile encounters an error, will not merge to new kubelet config")
+	} else {
+		oldKubeletConfig, ok := cfg.(*kubeletconfig.KubeletConfiguration)
+		if !ok {
+			klog.Warningf("getConfigFromFile gets an invalid kubelet config, will not merge to new kubelet config")
+		} else {
+			kubeletConfig.CgroupDriver = oldKubeletConfig.CgroupDriver
+			kubeletConfig.ResolverConfig = oldKubeletConfig.ResolverConfig
+		}
+	}
+
+	kubeletBytes, err := getConfigBytes(kubeletConfig)
+	if err != nil {
+		return err
+	}
+
+	return writeConfigBytesToDisk(kubeletBytes, kubeletDir)
+}
+
+// DownloadConfig downloads the kubelet configuration from a ConfigMap, re-build it with OS settings and writes it to disk.
 // Used at "kubeadm join" time
-func DownloadConfig(client clientset.Interface, kubeletVersion *version.Version, kubeletDir string) error {
+func DownloadConfig(client clientset.Interface, kubeletVersion *version.Version, nodeReg *kubeadmapi.NodeRegistrationOptions, kubeletDir string) error {
+
+	kubeletConfig, err := fetchConfigFromConfigMap(client, kubeletVersion, kubeletDir)
+	if err != nil {
+		return err
+	}
+
+	kubeletBytes, err := getConfigBytes(buildDynamicDefaults(kubeletConfig, nodeReg))
+	if err != nil {
+		return err
+	}
+
+	return writeConfigBytesToDisk(kubeletBytes, kubeletDir)
+}
+
+// fetchConfigFromConfigMap returns the kubelet configuration from a ConfigMap
+func fetchConfigFromConfigMap(client clientset.Interface, kubeletVersion *version.Version, kubeletDir string) (*kubeletconfig.KubeletConfiguration, error) {
 
 	// Download the ConfigMap from the cluster based on what version the kubelet is
 	configMapName := kubeadmconstants.GetKubeletConfigMapName(kubeletVersion)
 
 	fmt.Printf("[kubelet-start] Downloading configuration for the kubelet from the %q ConfigMap in the %s namespace\n",
 		configMapName, metav1.NamespaceSystem)
 
-	kubeletCfg, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(configMapName, metav1.GetOptions{})
-	// If the ConfigMap wasn't found and the kubelet version is v1.10.x, where we didn't support the config file yet
-	// just return, don't error out
-	if apierrors.IsNotFound(err) && kubeletVersion.Minor() == 10 {
-		return nil
+	cfg, err := componentconfigs.GetFromKubeletConfigMap(client, kubeletVersion)
+	if err != nil {
+		return nil, err
 	}
+
+	kubeletConfig, ok := cfg.(*kubeletconfig.KubeletConfiguration)
+	if !ok {
+		return nil, errors.New("fetchConfigFromConfigMap gets an invalid kubelet config")
```

Comment:
please change the error message just to:
`received an invalid KubeletConfiguration`